### ISSUE TITLE
[Support] Strange behavior unicodepwd and userAccountControl

### ISSUE THREAD

**Issue by nohnaimer** [DirectoryTree/LdapRecord#616]
**Environment:**
 - LDAP Server Type: [e.g. ActiveDirectory / OpenLDAP / FreeIPA] ActiveDirectory Windows Server 2019
 - PHP Version: [e.g. 8.1 / 8.2] 8.1
 - LdapRecord: 3.0.6

It used to work like this and everything was simple and clear:
```php
use LdapRecord\Models\ActiveDirectory\User;

$user = (new User)->inside('ou=Users,dc=local,dc=com');

$user->cn = 'John Doe';
$user->unicodePwd = 'SecretPassword';
$user->samaccountname = 'jdoe';
$user->userPrincipalName = 'jdoe@acme.org';

$user->save();

// Sync the created users attributes.
$user->refresh();

// Enable the user.
$user->userAccountControl = 512;

try {
    $user->save();
} catch (\LdapRecord\LdapRecordException $e) {
    // Failed saving user.
}
```
Now it just works like this:
```php
use LdapRecord\Models\ActiveDirectory\User;

$user = (new User)->inside('ou=Users,dc=local,dc=com');
$user->cn = 'John Doe';
$user->samaccountname = 'jdoe';
$user->userPrincipalName = 'jdoe@acme.org';

try {
      $user->save();
} catch (LdapRecordException $e) {}

$user->unicodepwd = ['', 'SecretPassword'];

try {
      $user->save();
} catch (LdapRecordException $e) {}

$user->refresh();
$user->update(['pwdlastset' => -1]);

$user->setFirstAttribute('userAccountControl', 512);

try {
      $user->save();
} catch (LdapRecordException $e) {}
```

Maybe I'm doing something wrong? Or can it be made even more difficult?

P.S. By the way, according to the documentation, it doesn't work - https://ldaprecord.com/docs/core/v3/active-directory/users/#creation

**Comment by stevebauman** [comment]
To follow up on my previous comment -- if that is an ill-intentioned comment, that's not something I'll tolerate on my repositories.

> So this is normal behavior? Or is it just a transitional period of version 3?

Nothing has been changed from v2 - v3 in regards to setting or changing passwords:

https://github.com/DirectoryTree/LdapRecord/commits/master/src/Models/Concerns/HasPassword.php

**Comment by nohnaimer** [comment]
No malice, just a question: if nothing has changed, then why doesn't it work?

[#504](https://github.com/DirectoryTree/LdapRecord/issues/504) I saw this issue bit by bit and put together a working version.

Just the further - the more save for each attribute. At the very beginning there were 2 saves, but now we need 3 saves and 1 update.

It would be nice to update the documentation. Because you take the code from there, but it does not really correspond - it does not work.

And returning to the user creation scheme - can it be simplified somehow? Now not all attributes are preserved as they were before...

**Comment by stevebauman** [comment]
> No malice, just a question: if nothing has changed, then why doesn't it work?

That's something we can get to the bottom of here 👍 

> It would be nice to update the documentation. Because you take the code from there, but it does not really correspond - it does not work.

If the documented method is no longer working, then let's get to the bottom of the issue.

Can you please try setting these attributes in a raw manner (bypassing the model's mutators) to see if this resolves your issue?

```php
use LdapRecord\Models\ActiveDirectory\User;

$user = (new User)->inside('ou=Users,dc=local,dc=com');

$user->cn = 'John Doe';
$user->samaccountname = 'jdoe';
$user->userPrincipalName = 'jdoe@acme.org';

$user->setRawAttribute('password', 'foo');
$user->setRawAttribute('userAccountControl', 512);

// User should be enabled and created with a password.
$user->save();
```

**Comment by nohnaimer** [comment]
Thanks, that works, but it's like going back to the stone age. The extension is used just because of the model layer...

In general, I’m ready to help with Active Directory debugging so that everything is beautiful again!

**Comment by stevebauman** [comment]
@nohnaimer I'm not saying the above example is a permanent fix -- I'm only asking to see if it resolves your issue so I can push an update to resolve the issue permanently 👍 

I'll push an update momentarily to resolve this 🙏

**Comment by stevebauman** [comment]
Hi @nohnaimer,

I just submitted a PR that should resolve this issue (https://github.com/DirectoryTree/LdapRecord/pull/617).

Can you target the `dev-master` branch of `ldaprecord` inside of your `composer.json` file, run `composer update` and try the below?

```php
use LdapRecord\Models\ActiveDirectory\User;

$user = (new User)->inside('ou=Users,dc=local,dc=com');

$user->cn = 'John Doe';
$user->unicodePwd = 'SecretPassword';
$user->userAccountControl = 512;
$user->samaccountname = 'jdoe';
$user->userPrincipalName = 'jdoe@acme.org';

try {
    $user->save();
} catch (\LdapRecord\LdapRecordException $e) {
    // Failed saving user.
}
```

**Comment by nohnaimer** [comment]
Hi @stevebauman,

This code works on TLS and SSL connection too:
```php
use LdapRecord\Models\ActiveDirectory\User;

$user = (new User)->inside('ou=Users,dc=local,dc=com');

$user->cn = 'John Doe';
$user->unicodePwd = 'SecretPassword';
$user->userAccountControl = 512;
$user->samaccountname = 'jdoe';
$user->userPrincipalName = 'jdoe@acme.org';

try {
    $user->save();
} catch (\LdapRecord\LdapRecordException $e) {
    // Failed saving user.
}
```
But instead attribute unicodePwd I use unicodepwd. Anyway it works and userAccountControl is set correctly. In production)

**Comment by stevebauman** [comment]
@nohnaimer Excellent, thanks for testing this out. I've just released [v3.1.0](https://github.com/DirectoryTree/LdapRecord/releases/tag/v3.1.0) with this patch. Run `composer update` and you should be all set 🙏
